pkg/template/apis/template/types.go

```diff
@@ -143,13 +162,52 @@ type TemplateInstanceConditionType string
 
 const (
 	// TemplateInstanceReady indicates the readiness of the template
-	// instantiation.
+	// instantiation or one of its objects.
 	TemplateInstanceReady TemplateInstanceConditionType = "Ready"
 	// TemplateInstanceInstantiateFailure indicates the failure of the template
-	// instantiation
+	// instantiation or one of its objects.
 	TemplateInstanceInstantiateFailure TemplateInstanceConditionType = "InstantiateFailure"
+	// TemplateInstanceWaiting indicates waiting for readiness or failure of the
+	// template instantiation or one of its objects.
+	TemplateInstanceWaiting TemplateInstanceConditionType = "Waiting"
 )
 
+// TemplateInstanceObject references an object created by a TemplateInstance.
+type TemplateInstanceObject struct {
+	// ref is a reference to the created object.  When used under .spec, only
+	// name and namespace are used; these can contain references to parameters
+	// which will be substituted following the usual rules.
+	Ref kapi.ObjectReference
+
+	// successRequirements hold all the requirements that must be met to
+	// consider the Template instantiation a success.  These requirements are
+	// ANDed together.
+	SuccessRequirements []TemplateCompletionRequirement
+
+	// failureRequirements hold the requirements that if any is met will cause
+	// the Template instantiation to be considered a failure.  These
+	// requirements are ORed together.
+	FailureRequirements []TemplateCompletionRequirement
+
+	// conditions represent the latest available observations of the object's
+	// current state.  This is not used in .spec.
+	Conditions []TemplateInstanceCondition
+}
+
+// TemplateCompletionRequirement holds a single requirement that is matched to
+// partially determine success or failure of a Template instantiation.
+type TemplateCompletionRequirement struct {
+	// jsonPath specifies a JSONPath expression which is run against an object.
+	JSONPath *string
+
+	// condition specifies the name of a condition to be looked up on an object.
+	Condition *string
+
+	// equals is the value which should be matched for the requirement to be
+	// fulfilled.
+	Equals string
```

Comment:
TargetValue?  ExpectedValue?  Equals is a bit weird since you're compounding the operation that will be performed on the field value, with the field that's holding the value.